### PR TITLE
Device streaming fixes

### PR DESCRIPTION
**Pull request by ewertons**
# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `master` branch. 
  - [ ] I have merged the latest `master` branch prior to submission and re-merged as needed after I took any feedback.
  - [ ] I have squashed my changes into one with a clear description of the change.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem
Fixes are needed to address:
https://github.com/Azure/azure-iot-sdk-c/issues/901
https://github.com/Azure/azure-iot-sdk-c/issues/900
https://github.com/Azure/azure-iot-sdk-c/issues/820

# Description of the solution
 
Implemented the changes needed.

### CODE REVIEW

File: iothub_client/devdoc/requirement_docs/iothub_client_streaming_requirements.md [hunk 1]
```diff
@@ -25,17 +25,17 @@ typedef struct DEVICE_STREAM_C2D_RESPONSE_TAG
...
 typedef DEVICE_STREAM_C2D_RESPONSE* (*DEVICE_STREAM_C2D_REQUEST_CALLBACK)(DEVICE_STREAM_C2D_REQUEST* request, const void* context);
 
-extern DEVICE_STREAM_C2D_RESPONSE* stream_c2d_response_create(DEVICE_STREAM_C2D_REQUEST* request, bool accept, char* data, char* content_type, char* content_encoding);
-extern DEVICE_STREAM_C2D_REQUEST* stream_c2d_request_clone(DEVICE_STREAM_C2D_REQUEST* request);
-extern void stream_c2d_response_destroy(DEVICE_STREAM_C2D_RESPONSE* response);
-extern void stream_c2d_request_destroy(DEVICE_STREAM_C2D_REQUEST* request);
+extern DEVICE_STREAM_C2D_RESPONSE* IoTHubClient_StreamC2DResponseCreate(DEVICE_STREAM_C2D_REQUEST* request, bool accept, char* data, char* content_type, char* content_encoding);
+extern DEVICE_STREAM_C2D_REQUEST* IoTHubClient_StreamC2DRequestClone(DEVICE_STREAM_C2D_REQUEST* request);
+extern void IoTHubClient_StreamC2DResponseDestroy(DEVICE_STREAM_C2D_RESPONSE* response);
+extern void IoTHubClient_StreamC2DRequestDestroy(DEVICE_STREAM_C2D_REQUEST* request);
```
Comment (mamokarz) [inline iothub_client/devdoc/requirement_docs/iothub_client_streaming_requirements.md]: Isn't it breaking change?

Comment (ewertons) [reply iothub_client/devdoc/requirement_docs/iothub_client_streaming_requirements.md]: Normally yes, but this is a public preview feature.

File: iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c [hunk 2]
```diff
@@ -244,15 +244,15 @@ static UWS_CLIENT_HANDLE create_websocket_client(DEVICE_STREAM_C2D_REQUEST* stre
     size_t host_length;
     const char* path;
     size_t path_length;
-    int port;
+    size_t port;
```
Comment (mamokarz) [inline iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c]: >size_t port; [](start = 4, length = 12)

Define port as size_t is odd. It should be uint16_t or uint32_t.

Comment (ewertons) [reply iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c]: size_t is uint32_t on 32 bit machines.

File: iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c [hunk 3]
```diff
@@ -244,15 +244,15 @@ static UWS_CLIENT_HANDLE create_websocket_client(DEVICE_STREAM_C2D_REQUEST* stre
...
 
     char host_address[1024];
     char resource_name[1024];
 
     ws_url = ws_url_create(stream_request->url);
```
Comment (mamokarz) [inline iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c]: >ws_url_create [](start = 13, length = 13)

There are some issues with these ws_url, not only the port is size_t but the gets for both host and path are guessing the buffer size.  Fix it does not need to be part of this PR but we should open a issue to fix it in the future.

File: iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c (hunk 2 above)
Comment (mamokarz) [reply iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c]: Sure, but the meaning is different. size_t is for size, this port is probably a number.

---
In reply to: [338800104](https://github.com/Azure/azure-iot-sdk-c/pull/1282#discussion_r338800104) [](ancestors = 338800104)

File: iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c (hunk 3 above)
Comment (ewertons) [reply iothub_client/samples/iothub_client_c2d_streaming_proxy_sample/iothub_client_c2d_streaming_proxy_sample.c]: Filed the issue to track it.
https://github.com/Azure/azure-c-shared-utility/issues/404

File: iothub_client/inc/iothub_client_streaming.h [hunk 4]
```diff
@@ -56,12 +56,12 @@ typedef struct DEVICE_STREAM_C2D_RESPONSE_TAG
...
 * @return    An instance of DEVICE_STREAM_C2D_RESPONSE indicating if the stream request is accepted or rejected.
 */
-typedef DEVICE_STREAM_C2D_RESPONSE* (*DEVICE_STREAM_C2D_REQUEST_CALLBACK)(DEVICE_STREAM_C2D_REQUEST* request, void* context);
+typedef DEVICE_STREAM_C2D_RESPONSE* (*DEVICE_STREAM_C2D_REQUEST_CALLBACK)(const DEVICE_STREAM_C2D_REQUEST* request, void* context);
 
-MOCKABLE_FUNCTION(, DEVICE_STREAM_C2D_RESPONSE*, stream_c2d_response_create, DEVICE_STREAM_C2D_REQUEST*, request, bool, accept);
-MOCKABLE_FUNCTION(, DEVICE_STREAM_C2D_REQUEST*, stream_c2d_request_clone, DEVICE_STREAM_C2D_REQUEST*, request);
-MOCKABLE_FUNCTION(, void, stream_c2d_response_destroy, DEVICE_STREAM_C2D_RESPONSE*, response);
-MOCKABLE_FUNCTION(, void, stream_c2d_request_destroy, DEVICE_STREAM_C2D_REQUEST*, request);
+MOCKABLE_FUNCTION(, DEVICE_STREAM_C2D_RESPONSE*, IoTHubClient_StreamC2DResponseCreate, const DEVICE_STREAM_C2D_REQUEST*, request, bool, accept);
```
Comment (danewalton) [inline iothub_client/inc/iothub_client_streaming.h]: >MOCKABLE_FUNCTION [](start = 0, length = 17)

nit: these should probably have some doxygen comments to go with them

**Pull request merged**